File: app/Sources/LLMClient.swift
import Foundation

struct LLMProvider {
    let endpoint: URL
    let model: String
    let apiKey: String
}

struct LLMClient {
    private static let timeout: TimeInterval = 15

    static func resolve() -> LLMProvider? {
        let defaults = UserDefaults.standard
        let provider = defaults.string(forKey: "provider") ?? "deepseek"

        switch provider {
        case "groq":
            guard let key = defaults.string(forKey: "groq_api_key"), !key.isEmpty else { return nil }
            let model = defaults.string(forKey: "groq_model") ?? "llama-3.3-70b-versatile"
            return LLMProvider(
                endpoint: URL(string: "https://api.groq.com/openai/v1/chat/completions")!,
                model: model,
                apiKey: key
            )
        case "custom":
            guard let key = defaults.string(forKey: "custom_api_key"), !key.isEmpty,
                  let ep = defaults.string(forKey: "custom_endpoint"), !ep.isEmpty,
                  let url = URL(string: ep),
                  let model = defaults.string(forKey: "custom_model"), !model.isEmpty
            else { return nil }
            return LLMProvider(endpoint: url, model: model, apiKey: key)
        default: // deepseek
            guard let key = defaults.string(forKey: "deepseek_api_key"), !key.isEmpty else { return nil }
            return LLMProvider(
                endpoint: URL(string: "https://api.deepseek.com/v1/chat/completions")!,
                model: "deepseek-chat",
                apiKey: key
            )
        }
    }

    static func chatCompletion(provider: LLMProvider, systemPrompt: String, userMessage: String) -> String? {
        let body: [String: Any] = [
            "model": provider.model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": userMessage],
            ],
            "temperature": 0.0,
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else { return nil }

        var request = URLRequest(url: provider.endpoint, timeoutInterval: timeout)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(provider.apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = jsonData

        let semaphore = DispatchSemaphore(value: 0)
        var result: String?

        URLSession.shared.dataTask(with: request) { data, _, error in
            defer { semaphore.signal() }
            guard error == nil, let data else { return }
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let choices = json["choices"] as? [[String: Any]],
                  let message = choices.first?["message"] as? [String: Any],
                  let content = message["content"] as? String else { return }
            result = content
        }.resume()

        semaphore.wait()
        return result
    }
}

File: linux/shuohua/llm_client.py
"""OpenAI-compatible HTTP client for LLM APIs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import httpx

from .config import Config
from .logger import slog

_TIMEOUT = 15.0


@dataclass
class LLMProvider:
    endpoint: str
    model: str
    api_key: str


def resolve(cfg: Config) -> Optional[LLMProvider]:
    """Resolve the configured LLM provider, or None if not configured."""
    provider = cfg.provider

    if provider == "dashscope":
        if not cfg.dashscope_api_key:
            return None
        return LLMProvider(
            endpoint="https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions",
            model="qwen-turbo",
            api_key=cfg.dashscope_api_key,
        )
    elif provider == "groq":
        if not cfg.groq_api_key:
            return None
        return LLMProvider(
            endpoint="https://api.groq.com/openai/v1/chat/completions",
            model=cfg.groq_model or "llama-3.3-70b-versatile",
            api_key=cfg.groq_api_key,
        )
    elif provider == "custom":
        if not all([cfg.custom_api_key, cfg.custom_endpoint, cfg.custom_model]):
            return None
        return LLMProvider(
            endpoint=cfg.custom_endpoint,
            model=cfg.custom_model,
            api_key=cfg.custom_api_key,
        )
    else:  # deepseek (default)
        if not cfg.deepseek_api_key:
            return None
        return LLMProvider(
            endpoint="https://api.deepseek.com/v1/chat/completions",
            model="deepseek-chat",
            api_key=cfg.deepseek_api_key,
        )


def chat_completion(
    provider: LLMProvider, system_prompt: str, user_message: str,
    proxy: str = "",
) -> Optional[str]:
    """Send a chat completion request. Returns the response content or None."""
    body = {
        "model": provider.model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_message},
        ],
        "temperature": 0.0,
    }

    try:
        resp = httpx.post(
            provider.endpoint,
            json=body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {provider.api_key}",
            },
            timeout=_TIMEOUT,
            proxy=proxy or None,
        )
        resp.raise_for_status()
        data = resp.json()
        return data["choices"][0]["message"]["content"]
    except Exception as e:
        slog(f"LLM API 错误: {e}")
        return None
